# DAY 4 认识pandas

## 数据的读取和查看

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
# 读取数据
import pandas as pd
data = pd.read_csv(r'/content/drive/MyDrive/data/data_day4.csv')

In [7]:
type(data) # 类

pandas.core.frame.DataFrame

方法 ：人.跳舞()
属性：人.年龄

In [ ]:
data.isnull() # 布尔矩阵显示缺失值，这个方法返回一个布尔矩阵，也是dataframe对象，其中True表示对应位置的值是缺失值，False表示对应位置的值不是缺失值。

Id  Home Ownership  Annual Income  Years in current job  Tax Liens  \
0     False           False          False                  True      False   
1     False           False          False                 False      False   
2     False           False          False                 False      False   
3     False           False          False                 False      False   
4     False           False          False                 False      False   
...     ...             ...            ...                   ...        ...   
7495  False           False          False                 False      False   
7496  False           False          False                 False      False   
7497  False           False          False                 False      False   
7498  False           False           True                  True      False   
7499  False           False           True                 False      False   

      Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                       False                    False                False   
1                       False                    False                False   
2                       False                    False                False   
3                       False                    False                False   
4                       False                    False                False   
...                       ...                      ...                  ...   
7495                    False                    False                False   
7496                    False                    False                False   
7497                    False                    False                False   
7498                    False                    False                False   
7499                    False                    False                False   

      Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                         False                          True         False   
1                         False                          True         False   
2                         False                          True         False   
3                         False                          True         False   
4                         False                          True         False   
...                         ...                           ...           ...   
7495                      False                          True         False   
7496                      False                         False         False   
7497                      False                          True         False   
7498                      False                          True         False   
7499                      False                          True         False   

      Purpose   Term  Current Loan Amount  Current Credit Balance  \
0       False  False                False                   False   
1       False  False                False                   False   
2       False  False                False                   False   
3       False  False                False                   False   
4       False  False                False                   False   
...       ...    ...                  ...                     ...   
7495    False  False                False                   False   
7496    False  False                False                   False   
7497    False  False                False                   False   
7498    False  False                False                   False   
7499    False  False                False                   False   

      Monthly Debt  Credit Score  Credit Default  
0            False         False           False  
1            False         False           False  
2            False         False           False  
3            False         False           False  
4            False         False           False  
...            ...         

In [ ]:
data.head(10)

Id Home Ownership  Annual Income Years in current job  Tax Liens  \
0   0       Own Home       482087.0                  NaN        0.0   
1   1       Own Home      1025487.0            10+ years        0.0   
2   2  Home Mortgage       751412.0              8 years        0.0   
3   3       Own Home       805068.0              6 years        0.0   
4   4           Rent       776264.0              8 years        0.0   
5   5           Rent            NaN              7 years        0.0   
6   6  Home Mortgage      1511108.0            10+ years        0.0   
7   7           Rent      1040060.0            10+ years        0.0   
8   8  Home Mortgage            NaN              5 years        0.0   
9   9  Home Mortgage            NaN               1 year        0.0   

   Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                     11.0                     26.3             685960.0   
1                     15.0                     15.3            1181730.0   
2                     11.0                     35.0            1182434.0   
3                      8.0                     22.5             147400.0   
4                     13.0                     13.6             385836.0   
5                     12.0                     14.6             366784.0   
6                      9.0                     20.3             388124.0   
7                     13.0                     12.0             330374.0   
8                     17.0                     15.7                  0.0   
9                     10.0                     24.6             511302.0   

   Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                        1.0                           NaN           1.0   
1                        0.0                           NaN           0.0   
2                        0.0                           NaN           0.0   
3                        1.0                           NaN           1.0   
4                        1.0                           NaN           0.0   
5                        0.0                           NaN           0.0   
6                        0.0                          73.0           0.0   
7                        0.0                          18.0           0.0   
8                        1.0                           NaN           1.0   
9                        0.0                           6.0           0.0   

              Purpose        Term  Current Loan Amount  \
0  debt consolidation  Short Term           99999999.0   
1  debt consolidation   Long Term             264968.0   
2  debt consolidation  Short Term           99999999.0   
3  debt consolidation  Short Term             121396.0   
4  debt consolidation  Short Term             125840.0   
5               other   Long Term             337304.0   
6   home improvements  Short Term           99999999.0   
7               other  Short Term             250888.0   
8   home improvements  Short Term             129734.0   
9  debt consolidation   Long Term             572880.0   

   Current Credit Balance  Monthly Debt  Credit Score  Credit Default  
0                 47386.0        7914.0         749.0               0  
1                394972.0       18373.0         737.0               1  
2                308389.0       13651.0         742.0               0  
3                 95855.0       11338.0         694.0               0  
4                 93309.0        7180.0         719.0               0  
5                165680.0       18692.0           NaN               1  
6                 51623.0        2317.0         745.0               0  
7                 89015.0       19761.0         705.0               1  
8                    19.0          17.0           NaN               0  
9                205333.0       17613.0           NaN               1

In [8]:
# pip install openpyxl
# pandas读取excel需要安装openpyxl库，去anaconda prompt中安装
data2 =pd.read_excel('data.xlsx')
data2

FileNotFoundError: [Errno 2] No such file or directory: 'data.xlsx'

In [ ]:
data.head(10)

Id Home Ownership  Annual Income Years in current job  Tax Liens  \
0   0       Own Home       482087.0                  NaN        0.0   
1   1       Own Home      1025487.0            10+ years        0.0   
2   2  Home Mortgage       751412.0              8 years        0.0   
3   3       Own Home       805068.0              6 years        0.0   
4   4           Rent       776264.0              8 years        0.0   
5   5           Rent            NaN              7 years        0.0   
6   6  Home Mortgage      1511108.0            10+ years        0.0   
7   7           Rent      1040060.0            10+ years        0.0   
8   8  Home Mortgage            NaN              5 years        0.0   
9   9  Home Mortgage            NaN               1 year        0.0   

   Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                     11.0                     26.3             685960.0   
1                     15.0                     15.3            1181730.0   
2                     11.0                     35.0            1182434.0   
3                      8.0                     22.5             147400.0   
4                     13.0                     13.6             385836.0   
5                     12.0                     14.6             366784.0   
6                      9.0                     20.3             388124.0   
7                     13.0                     12.0             330374.0   
8                     17.0                     15.7                  0.0   
9                     10.0                     24.6             511302.0   

   Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                        1.0                           NaN           1.0   
1                        0.0                           NaN           0.0   
2                        0.0                           NaN           0.0   
3                        1.0                           NaN           1.0   
4                        1.0                           NaN           0.0   
5                        0.0                           NaN           0.0   
6                        0.0                          73.0           0.0   
7                        0.0                          18.0           0.0   
8                        1.0                           NaN           1.0   
9                        0.0                           6.0           0.0   

              Purpose        Term  Current Loan Amount  \
0  debt consolidation  Short Term           99999999.0   
1  debt consolidation   Long Term             264968.0   
2  debt consolidation  Short Term           99999999.0   
3  debt consolidation  Short Term             121396.0   
4  debt consolidation  Short Term             125840.0   
5               other   Long Term             337304.0   
6   home improvements  Short Term           99999999.0   
7               other  Short Term             250888.0   
8   home improvements  Short Term             129734.0   
9  debt consolidation   Long Term             572880.0   

   Current Credit Balance  Monthly Debt  Credit Score  Credit Default  
0                 47386.0        7914.0         749.0               0  
1                394972.0       18373.0         737.0               1  
2                308389.0       13651.0         742.0               0  
3                 95855.0       11338.0         694.0               0  
4                 93309.0        7180.0         719.0               0  
5                165680.0       18692.0           NaN               1  
6                 51623.0        2317.0         745.0               0  
7                 89015.0       19761.0         705.0               1  
8                    19.0          17.0           NaN               0  
9                205333.0       17613.0           NaN               1

## 数据信息的查看

此时data是一个dataframe类型的对象，可以理解为dataframe类的实例。实例就具有类的属性和方法。
1. 属性的调用格式为：实例名.属性名。
2. 方法的调用格式为：实例名.方法名()。



In [ ]:
data.info()       # 列名、非空值、数据类型

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 18 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Id                            7500 non-null   int64  
 1   Home Ownership                7500 non-null   object 
 2   Annual Income                 5943 non-null   float64
 3   Years in current job          7129 non-null   object 
 4   Tax Liens                     7500 non-null   float64
 5   Number of Open Accounts       7500 non-null   float64
 6   Years of Credit History       7500 non-null   float64
 7   Maximum Open Credit           7500 non-null   float64
 8   Number of Credit Problems     7500 non-null   float64
 9   Months since last delinquent  3419 non-null   float64
 10  Bankruptcies                  7486 non-null   float64
 11  Purpose                       7500 non-null   object 
 12  Term                          7500 non-null   object 
 13  Cur

In [ ]:
data.shape        # (行数, 列数) data的属性

(7500, 18)

In [ ]:
data.columns      # 所有列名 data的属性

Index(['Id', 'Home Ownership', 'Annual Income', 'Years in current job',
       'Tax Liens', 'Number of Open Accounts', 'Years of Credit History',
       'Maximum Open Credit', 'Number of Credit Problems',
       'Months since last delinquent', 'Bankruptcies', 'Purpose', 'Term',
       'Current Loan Amount', 'Current Credit Balance', 'Monthly Debt',
       'Credit Score', 'Credit Default'],
      dtype='object')

In [ ]:
data.describe()   # 数值列的基本统计量

Id  Annual Income    Tax Liens  Number of Open Accounts  \
count  7500.000000   5.943000e+03  7500.000000              7500.000000   
mean   3749.500000   1.366392e+06     0.030133                11.130933   
std    2165.207842   8.453392e+05     0.271604                 4.908924   
min       0.000000   1.645970e+05     0.000000                 2.000000   
25%    1874.750000   8.443410e+05     0.000000                 8.000000   
50%    3749.500000   1.168386e+06     0.000000                10.000000   
75%    5624.250000   1.640137e+06     0.000000                14.000000   
max    7499.000000   1.014934e+07     7.000000                43.000000   

       Years of Credit History  Maximum Open Credit  \
count              7500.000000         7.500000e+03   
mean                 18.317467         9.451537e+05   
std                   7.041946         1.602622e+07   
min                   4.000000         0.000000e+00   
25%                  13.500000         2.792295e+05   
50%                  17.000000         4.781590e+05   
75%                  21.800000         7.935015e+05   
max                  57.700000         1.304726e+09   

       Number of Credit Problems  Months since last delinquent  Bankruptcies  \
count                7500.000000                   3419.000000   7486.000000   
mean                    0.170000                     34.692600      0.117152   
std                     0.498598                     21.688806      0.347192   
min                     0.000000                      0.000000      0.000000   
25%                     0.000000                     16.000000      0.000000   
50%                     0.000000                     32.000000      0.000000   
75%                     0.000000                     50.000000      0.000000   
max                     7.000000                    118.000000      4.000000   

       Current Loan Amount  Current Credit Balance   Monthly Debt  \
count         7.500000e+03            7.500000e+03    7500.000000   
mean          1.187318e+07            2.898332e+05   18314.454133   
std           3.192612e+07            3.178714e+05   11926.764673   
min           1.124200e+04            0.000000e+00       0.000000   
25%           1.801690e+05            1.142565e+05   10067.500000   
50%           3.095730e+05            2.093230e+05   16076.500000   
75%           5.198820e+05            3.604062e+05   23818.000000   
max           1.000000e+08            6.506797e+06  136679.000000   

       Credit Score  Credit Default  
count   5943.000000     7500.000000  
mean    1151.087498        0.281733  
std     1604.451418        0.449874  
min      585.000000        0.000000  
25%      711.000000        0.000000  
50%      731.000000        0.000000  
75%      743.000000        1.000000  
max     7510.000000        1.000000

In [10]:
# dtype是data type的缩写，用于描述数据类型。后续会频繁借助这个方法来查看某一列数据的属性
data.dtypes       # 各列数据类型
# data.info()

Id                                int64
Home Ownership                   object
Annual Income                   float64
Years in current job             object
Tax Liens                       float64
Number of Open Accounts         float64
Years of Credit History         float64
Maximum Open Credit             float64
Number of Credit Problems       float64
Months since last delinquent    float64
Bankruptcies                    float64
Purpose                          object
Term                             object
Current Loan Amount             float64
Current Credit Balance          float64
Monthly Debt                    float64
Credit Score                    float64
Credit Default                    int64
dtype: object

In [ ]:
data["Annual Income"].dtype # 查看某一列的数据类型

dtype('float64')

In [ ]:
data.isnull()            # 布尔矩阵显示缺失值，这个方法返回一个布尔矩阵，也是dataframe对象，其中True表示对应位置的值是缺失值，False表示对应位置的值不是缺失值。

Id  Home Ownership  Annual Income  Years in current job  Tax Liens  \
0     False           False          False                  True      False   
1     False           False          False                 False      False   
2     False           False          False                 False      False   
3     False           False          False                 False      False   
4     False           False          False                 False      False   
...     ...             ...            ...                   ...        ...   
7495  False           False          False                 False      False   
7496  False           False          False                 False      False   
7497  False           False          False                 False      False   
7498  False           False           True                  True      False   
7499  False           False           True                 False      False   

      Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                       False                    False                False   
1                       False                    False                False   
2                       False                    False                False   
3                       False                    False                False   
4                       False                    False                False   
...                       ...                      ...                  ...   
7495                    False                    False                False   
7496                    False                    False                False   
7497                    False                    False                False   
7498                    False                    False                False   
7499                    False                    False                False   

      Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                         False                          True         False   
1                         False                          True         False   
2                         False                          True         False   
3                         False                          True         False   
4                         False                          True         False   
...                         ...                           ...           ...   
7495                      False                          True         False   
7496                      False                         False         False   
7497                      False                          True         False   
7498                      False                          True         False   
7499                      False                          True         False   

      Purpose   Term  Current Loan Amount  Current Credit Balance  \
0       False  False                False                   False   
1       False  False                False                   False   
2       False  False                False                   False   
3       False  False                False                   False   
4       False  False                False                   False   
...       ...    ...                  ...                     ...   
7495    False  False                False                   False   
7496    False  False                False                   False   
7497    False  False                False                   False   
7498    False  False                False                   False   
7499    False  False                False                   False   

      Monthly Debt  Credit Score  Credit Default  
0            False         False           False  
1            False         False           False  
2            False         False           False  
3            False         False           False  
4            False         False           False  
...            ...         

In [ ]:
type(data.isnull())         # 布尔矩阵显示缺失值，这个方法返回一个布尔矩阵，其中True表示对应位置的值是缺失值，False表示对应位置的值不是缺失值。

pandas.core.frame.DataFrame

In [ ]:
data.isnull()            # 布尔矩阵显示缺失值，这个方法返回一个布尔矩阵，其中True表示对应位置的值是缺失值，False表示对应位置的值不是缺失值。

Id  Home Ownership  Annual Income  Years in current job  Tax Liens  \
0     False           False          False                  True      False   
1     False           False          False                 False      False   
2     False           False          False                 False      False   
3     False           False          False                 False      False   
4     False           False          False                 False      False   
...     ...             ...            ...                   ...        ...   
7495  False           False          False                 False      False   
7496  False           False          False                 False      False   
7497  False           False          False                 False      False   
7498  False           False           True                  True      False   
7499  False           False           True                 False      False   

      Number of Open Accounts  Years of Credit History  Maximum Open Credit  \
0                       False                    False                False   
1                       False                    False                False   
2                       False                    False                False   
3                       False                    False                False   
4                       False                    False                False   
...                       ...                      ...                  ...   
7495                    False                    False                False   
7496                    False                    False                False   
7497                    False                    False                False   
7498                    False                    False                False   
7499                    False                    False                False   

      Number of Credit Problems  Months since last delinquent  Bankruptcies  \
0                         False                          True         False   
1                         False                          True         False   
2                         False                          True         False   
3                         False                          True         False   
4                         False                          True         False   
...                         ...                           ...           ...   
7495                      False                          True         False   
7496                      False                         False         False   
7497                      False                          True         False   
7498                      False                          True         False   
7499                      False                          True         False   

      Purpose   Term  Current Loan Amount  Current Credit Balance  \
0       False  False                False                   False   
1       False  False                False                   False   
2       False  False                False                   False   
3       False  False                False                   False   
4       False  False                False                   False   
...       ...    ...                  ...                     ...   
7495    False  False                False                   False   
7496    False  False                False                   False   
7497    False  False                False                   False   
7498    False  False                False                   False   
7499    False  False                False                   False   

      Monthly Debt  Credit Score  Credit Default  
0            False         False           False  
1            False         False           False  
2            False         False           False  
3            False         False           False  
4            False         False           False  
...            ...         

In [ ]:
data.isnull().sum()      # 每列缺失值计数,sum方法为求每一列的和

Id                                 0
Home Ownership                     0
Annual Income                   1557
Years in current job             371
Tax Liens                          0
Number of Open Accounts            0
Years of Credit History            0
Maximum Open Credit                0
Number of Credit Problems          0
Months since last delinquent    4081
Bankruptcies                      14
Purpose                            0
Term                               0
Current Loan Amount                0
Current Credit Balance             0
Monthly Debt                       0
Credit Score                    1557
Credit Default                     0
dtype: int64

## 缺失值的填补
尝试用众数 中位数进行填补

In [ ]:
data['Annual Income']

0        482087.0
1       1025487.0
2        751412.0
3        805068.0
4        776264.0
          ...    
7495     402192.0
7496    1533984.0
7497    1878910.0
7498          NaN
7499          NaN
Name: Annual Income, Length: 7500, dtype: float64

In [ ]:
type(data['Annual Income'])
# dataframe里单独的一列是series

pandas.core.series.Series

In [ ]:
# 计算 'Annual Income' 列的中位数（会自动忽略 NaN 值）
median_income = data['Annual Income'].median()
median_income

1168386.0

In [ ]:
# 使用计算出的中位数填补该列的 NaN 值
# inplace=True 参数表示直接在原 DataFrame 上进行修改
# 如果不设置该参数，fillna() 方法会返回一个新的 DataFrame，原 DataFrame 不会被修改
data['Annual Income'].fillna(median_income, inplace=True)

In [ ]:
# 检查下是否有缺失值
data['Annual Income'].isnull().sum()

0

In [ ]:
# 使用众数填充缺失值
import pandas as pd
data = pd.read_csv('data.csv') #需要重新读取一遍数据
mode = data['Annual Income'].mode()
# mode() 会返回数据中出现频率最高的所有值，如果频次相同，会返回最多每个值。
mode

0     969475.0
1    1043651.0
2    1058376.0
3    1161660.0
4    1338113.0
Name: Annual Income, dtype: float64

In [ ]:
# 这里返回了4个最多频次的值，我们一般保留第一个
mode = mode[0]

In [ ]:
# 众数填补
data['Annual Income'].fillna(mode, inplace=True)
# 检查下是否有缺失值
data['Annual Income'].isnull().sum()

0

## 填补所有的数值型缺失值
从上面的打印信息可以看到，缺失值的数值列包括：Annual Income、Months since last delinquent、Bankruptcies、Credit Score。这些列的缺失值数量分别为1557、4081、14和1557。

In [ ]:
data.columns

Index(['Id', 'Home Ownership', 'Annual Income', 'Years in current job',
       'Tax Liens', 'Number of Open Accounts', 'Years of Credit History',
       'Maximum Open Credit', 'Number of Credit Problems',
       'Months since last delinquent', 'Bankruptcies', 'Purpose', 'Term',
       'Current Loan Amount', 'Current Credit Balance', 'Monthly Debt',
       'Credit Score', 'Credit Default'],
      dtype='object')

In [ ]:
type(data.columns)

pandas.core.indexes.base.Index

很自然的 我们想要遍历打印这个对象中的每一个元素，如果是数值，就补全缺失值

In [ ]:
# 介绍一下tolist方法，将numpy数组和pandas对象转换成list
import numpy as np
a =np.array([1,2,3])
a.tolist()

[1, 2, 3]

In [12]:
c = data.columns.tolist()
type(c)

list

In [19]:
# 循环遍历c这个列表中的每一列，将数值列用均值填充
for i in c:
    # 找到为数值型的列
    # data.info()
    # print(data[i].dtype)
    if data[i].dtype != 'object': # 找到为数值型的列
        if data[i].isnull().sum() > 0: # 找到存在缺失值的列
            #计算该列的均值
            mean_value = data[i].mean()
            #用均值填充缺失值
            data[i].fillna(mean_value, inplace=True)

data.isnull().sum()

Id                                0
Home Ownership                    0
Annual Income                     0
Years in current job            371
Tax Liens                         0
Number of Open Accounts           0
Years of Credit History           0
Maximum Open Credit               0
Number of Credit Problems         0
Months since last delinquent      0
Bankruptcies                      0
Purpose                           0
Term                              0
Current Loan Amount               0
Current Credit Balance            0
Monthly Debt                      0
Credit Score                      0
Credit Default                    0
dtype: int64